File: 133-clone-graph/clone-graph.ts
function cloneGraph(node: _Node | null): _Node | null {
    const visited = new Map<_Node, _Node>();
    return clone(node, visited);
};

function clone(node: _Node | null, visited: Map<_Node, _Node>): _Node | null {
    if(node === null){
        return node;
    }

    if(visited.has(node)){
        return visited.get(node);
    }

    const copy = new Node(node.val);
    visited.set(node, copy);

    for(const neighbor of node.neighbors){
        copy.neighbors.push(clone(neighbor, visited));
    }
    
    return copy;
}

File: 133-clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None

        visited = {}

        def dfs(node):
            clone = Node(node.val)
            visited[node] = clone

            for neighbor in node.neighbors:
                if neighbor not in visited:
                    visited[node].neighbors.append(dfs(neighbor))
                else:
                    visited[node].neighbors.append(visited[neighbor])

            return clone

        return dfs(node)

        